apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -47,95 +68,261 @@
 %% FIXME
 -define(DS_SHARD, <<"local">>).
 
--define(WHEN_ENABLED(DO),
-    case is_store_enabled() of
-        true -> DO;
-        false -> {skipped, disabled}
-    end
-).
+-record(sessionds, {
+    %% Client ID
+    id :: binary(),
+    %% Client’s Subscriptions.
+    subscriptions :: map(),
+    iterators :: map(),
+    %%
+    conf
+}).
 
-%%
+-type session() :: #sessionds{}.
 
-init() ->
-    ?WHEN_ENABLED(begin
-        ok = emqx_ds:ensure_shard(?DS_SHARD, #{
-            dir => filename:join([emqx:data_dir(), ds, messages, ?DS_SHARD])
-        }),
-        ok = emqx_persistent_session_ds_router:init_tables(),
-        ok
-    end).
+-type clientinfo() :: emqx_types:clientinfo().
+-type conninfo() :: emqx_types:conninfo().
+-type replies() :: emqx_session:replies().
 
 %%
 
--spec persist_message(emqx_types:message()) ->
-    ok | {skipped, _Reason} | {error, _TODO}.
-persist_message(Msg) ->
-    ?WHEN_ENABLED(
-        case needs_persistence(Msg) andalso has_subscribers(Msg) of
-            true ->
-                store_message(Msg);
-            false ->
-                {skipped, needs_no_persistence}
-        end
-    ).
-
-needs_persistence(Msg) ->
-    not (emqx_message:get_flag(dup, Msg) orelse emqx_message:is_sys(Msg)).
-
-store_message(Msg) ->
-    ID = emqx_message:id(Msg),
-    Timestamp = emqx_guid:timestamp(ID),
-    Topic = emqx_topic:words(emqx_message:topic(Msg)),
-    emqx_ds_storage_layer:store(?DS_SHARD, ID, Timestamp, Topic, serialize_message(Msg)).
-
-has_subscribers(#message{topic = Topic}) ->
-    emqx_persistent_session_ds_router:has_any_route(Topic).
+-spec create(clientinfo(), conninfo(), emqx_session:conf()) ->
+    session().
+create(#{clientid := ClientID}, _ConnInfo, Conf) ->
+    #sessionds{
+        id = ClientID,
+        subscriptions = #{},
+        conf = Conf
+    }.
+
+-spec open(clientinfo(), conninfo()) ->
+    {true, session()} | false.
+open(#{clientid := ClientID}, _ConnInfo) ->
+    open_session(ClientID).
```

Comment:
Upon closer inspection, I think the PS implementation will need to do something similar to `emqx_cm:pick_channel` (perhaps calling `kick_session/2` for the living client and `discard_session/2` for the potentially stale ones would be more precise) to kick any living client during takeover.

Reply:
Very good point.